### ISSUE TITLE
Migration failed 06\01_JobMigration

### ISSUE THREAD

**Issue by vegar** [sergeyzwezdin/Hangfire.Mongo#122]
I'm trying to upgrade to latest version. Was stuck at the <0.4 release, but now that migrations are in its place, I though I should give it a try. 

I've gotten the latest libraries:
```
-  <package id="Hangfire.Core" version="1.6.8" targetFramework="net452" />
-  <package id="Hangfire.Mongo" version="0.3.2" targetFramework="net452" />
+  <package id="Hangfire.Core" version="1.6.17" targetFramework="net452" />
+  <package id="Hangfire.Mongo" version="0.5.7" targetFramework="net452" />
 
   <package id="MongoDB.Bson" version="2.5.0" targetFramework="net452" />
   <package id="MongoDB.Driver" version="2.5.0" targetFramework="net452" />
   <package id="MongoDB.Driver.Core" version="2.5.0" targetFramework="net452" />
```
And changed my configuration to migrate data:

```
GlobalConfiguration.Configuration.UseMongoStorage(mongoUrl.ToString(), mongoUrl.DatabaseName, new MongoStorageOptions
{
  MigrationOptions = new MongoMigrationOptions
  {
    BackupStrategy = MongoBackupStrategy.Collections,
    Strategy = MongoMigrationStrategy.Migrate
  }
});
```

At startup I get an exception

```
Microsoft.WindowsAzure.ServiceRuntime Critical: 31337 : System.InvalidCastException: Unable to cast object of type 'MongoDB.Bson.BsonDocument' to type 'MongoDB.Bson.BsonString'.
   at MongoDB.Bson.BsonValue.get_AsString()
   at Hangfire.Mongo.Migration.Steps.Version06.JobMigration.<>c.<Execute>b__4_6(BsonDocument s)
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at MongoDB.Bson.BsonArray.AddRange(IEnumerable`1 values)
   at Hangfire.Mongo.Migration.Steps.Version06.JobMigration.<>c__DisplayClass4_0.<Execute>b__3(BsonDocument job)

```

By stepping through, I've identified the following line being the cause:
https://github.com/sergeyzwezdin/Hangfire.Mongo/blob/671465941b21bd74671d3ce556094db70a6a333e/src/Hangfire.Mongo/Migration/Steps/Version06/01_JobMigration.cs#L43-L49

More specific, it's the `s["Data"].ToString`-part that throws:

![image](https://user-images.githubusercontent.com/764517/34295694-faab3fcc-e70e-11e7-8263-f21df65abb11.png)

The defined states for the job it is trying to migrate is as follows. This is the first job that it tries to migrate, and it throws at the first state:

```
{ 
    "_id" : ObjectId("5a3b6836faf8a74d34f1430f"), 
    "JobId" : NumberInt(56), 
    "Name" : "Enqueued", 
    "Reason" : null, 
    "CreatedAt" : ISODate("2017-12-21T07:52:22.252+0000"), 
    "Data" : {
        "EnqueuedAt" : "2017-12-21T07:52:19.8959066Z", 
        "Queue" : "default"
    }
}
{ 
    "_id" : ObjectId("5a3b6844faf8a74d34f14310"), 
    "JobId" : NumberInt(56), 
    "Name" : "Processing", 
    "Reason" : null, 
    "CreatedAt" : ISODate("2017-12-21T07:52:36.246+0000"), 
    "Data" : {
        "StartedAt" : "2017-12-21T07:52:36.1371576Z", 
        "ServerId" : "no-fin-sc-0008@v1.0.0.0:19764:6c38373d-4a8f-4c3b-975d-b00181ddb57c", 
        "WorkerId" : "f02d59fa-63e2-426a-8d40-152c283efcfb"
    }
}
{ 
    "_id" : ObjectId("5a3b6844faf8a74d34f14313"), 
    "JobId" : NumberInt(56), 
    "Name" : "Succeeded", 
    "Reason" : null, 
    "CreatedAt" : ISODate("2017-12-21T07:52:36.901+0000"), 
    "Data" : {
        "SucceededAt" : "2017-12-21T07:52:36.8831964Z", 
        "PerformanceDuration" : "611", 
        "Latency" : "16123"
    }
}

```

**Comment by MarLoe** [comment]
@vegar thank you for your thorough report.
I'll take a look at this and see why we expect a string when in fact you got a document.

**Comment by MarLoe** [comment]
@vegar , I am a bit puzzled here. Normally I would expect the "Data" field to be a raw JSON string at this point.

The easy fix would be to check the type of the "Data" field before doing an ".AsString" on it.
However I would like to try and understand why we ended up in this situation.

Is there any way you could export your mongo hangfire database and make it available to me?

**Comment by vegar** [comment]
Doesn't look like the hangfire tables contains anything worth hiding, so here you go:

[hangfire.mongo.dump.zip](https://github.com/sergeyzwezdin/Hangfire.Mongo/files/1599633/hangfire.mongo.dump.zip)

It's a zipped up `mongodump` and should be restored with `mongorestore` after unzipping to a folder.

**Comment by vegar** [comment]
> I would expect the `Data` field to be a raw JSON string at this point.

And from what I can see, it is for some documents:
![image](https://user-images.githubusercontent.com/764517/34511953-5358dcb6-f05f-11e7-8cc0-6257f7a3a2fe.png)

**Comment by MarLoe** [comment]
@vegar, I have downloaded and imported you dump successfully - you can remove the link if you fell like.

Has this database been through any migrations before it came to this state?

I am asking because I am clueless why some of the "Data" fields has changed from raw JSON strings to actual objects.

Long story short, I am using you dump to incorporate a workaround so the migration step should work despite this quirk.

**Comment by vegar** [comment]
Not as far as I know.

I use Studio 3T, and their Schema analytics reports the following on our production database:

```
> Analyzed all 20003 documents
> Inspecting filed 'Data'
>    Type distribution
>          Object: 90.3%
>          String: 9.7%

```
By closer inspection, all documents containing string type, is created before 2016-07-01. All documents after that date contains objects.

**Comment by MarLoe** [comment]
@vegar, I give up understanding this. I have added improved migration to take this case into account. It seem to work. It sits on the 0.5.8 branch. You can try it out, and give feedback if tit does the trick.

**Comment by vegar** [comment]
As far as I can see, the migration worked this time. 
The schema collection contains a document with the `_id` 11.

The backed up collections are not deleted after migration, but I guess that's by design?

**Comment by MarLoe** [comment]
@vegar, super - good to hear. 

The left behind backed-up collections is by design. Should it turn out that moving on to the next version reveals issues at a later time, you are still able to (manually) restore to previous version.

Perhaps a feature request is a scheduled job that cleans out the backup collections after 30 days ;)

**Comment by MarLoe** [comment]
Fixed in 0.5.8
